sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala

```diff
@@ -564,4 +566,28 @@ class FileMetadataStructSuite extends QueryTest with SharedSparkSession {
       )
     }
   }
+
+  Seq(true, false).foreach { useVectorizedReader =>
+    val label = if (useVectorizedReader) "reading batches" else "reading rows"
+    test(s"SPARK-39806: metadata for a partitioned table ($label)") {
+      withSQLConf(SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key -> useVectorizedReader.toString) {
+        withTempPath { dir =>
+          // Store dynamically partitioned data.
+          Seq(1 -> 1).toDF("a", "b").write.format("parquet").partitionBy("b")
+            .save(dir.getAbsolutePath)
+
+          // Identify the data file and its metadata.
+          // We expect there to be exactly one subdirectory containing exactly one parquet file.
+          val subdirectory = dir.listFiles().filter(_.isDirectory).head
```

Comment:
If the number of files is not too large, maybe we can change to use 
```
val file = TestUtils.listDirectory(dir).head
```

Reply:
Ok, switched to `TestUtils.recursiveList` since it returns `File`s instead of `String` and fits here a little bit better.